frame/nomination-pools/src/lib.rs

```diff
@@ -0,0 +1,1967 @@
+		*who == self.roles.root || *who == self.roles.state_toggler
+	}
+
+	fn is_destroying(&self) -> bool {
+		matches!(self.state, PoolState::Destroying)
+	}
+
+	/// Whether or not the pool is ok to be in `PoolSate::Open`. If this returns an `Err`, then the
+	/// pool is unrecoverable and should be in the destroying state.
+	fn ok_to_be_open(&self) -> Result<(), DispatchError> {
+		ensure!(!self.is_destroying(), Error::<T>::CanNotChangeState);
+
+		let bonded_balance =
+			T::StakingInterface::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
+		ensure!(!bonded_balance.is_zero(), Error::<T>::OverflowRisk);
+
+		let points_to_balance_ratio_floor = self
+			.points
+			// We checked for zero above
+			.div(bonded_balance);
+
+		// Pool points can inflate relative to balance, but only if the pool is slashed.
+		// If we cap the ratio of points:balance so one cannot join a pool that has been slashed
+		// 90%,
+		ensure!(points_to_balance_ratio_floor < 10u32.into(), Error::<T>::OverflowRisk);
+		// while restricting the balance to 1/10th of max total issuance,
+		ensure!(
+			bonded_balance < BalanceOf::<T>::max_value().div(10u32.into()),
+			Error::<T>::OverflowRisk
+		);
+		// then we can be decently confident the bonding pool points will not overflow
+		// `BalanceOf<T>`. Note that these are just heuristics.
+
+		Ok(())
+	}
+
+	/// Check that the pool can accept a member with `new_funds`.
+	fn ok_to_join(&self) -> Result<(), DispatchError> {
+		ensure!(self.state == PoolState::Open, Error::<T>::NotOpen);
+		self.ok_to_be_open()?;
+		Ok(())
+	}
+
+	fn ok_to_unbond_other_with(
+		&self,
+		caller: &T::AccountId,
+		target_account: &T::AccountId,
+		target_delegator: &Delegator<T>,
+	) -> Result<(), DispatchError> {
+		let is_permissioned = caller == target_account;
+		let is_depositor = *target_account == self.roles.depositor;
+		match (is_permissioned, is_depositor) {
+			// If the pool is blocked, then an admin with kicking permissions can remove a
+			// delegator. If the pool is being destroyed, anyone can remove a delegator
+			(false, false) => {
+				ensure!(
+					self.can_kick(caller) || self.is_destroying(),
+					Error::<T>::NotKickerOrDestroying
+				)
+			},
+			// Any delegator who is not the depositor can always unbond themselves
+			(true, false) => (),
+			// The depositor can only start unbonding if the pool is already being destroyed and
+			// they are the delegator in the pool. Note that an invariant is once the pool is
+			// destroying it cannot switch states, so by being in destroying we are guaranteed no
+			// other delegators can possibly join.
+			(false, true) | (true, true) => {
+				ensure!(target_delegator.points == self.points, Error::<T>::NotOnlyDelegator);
+				ensure!(self.is_destroying(), Error::<T>::NotDestroying);
+			},
+		};
+		Ok(())
+	}
+
+	/// Returns a result indicating if [`Pallet::withdraw_unbonded_other`] can be executed.
+	fn ok_to_withdraw_unbonded_other_with(
```

Comment:
It does seem to me like semantically `ok_to_withdraw_unbonded_other_with` should have been a function defined in the `impl` of `SubPools` rather than this.. :curious:

Reply:
Its a bit of a toss up because we still need multiple pieces of permissioning info from the bonded pool. I think my logic was that the sub pools and reward pool conceptually extend the bonded pool (not saying this is great justification though)